tests/math/test_functions.py

```diff
@@ -2478,3 +2478,14 @@ def test_jax_ndim():
         _ = qml.math.ndim(jax.numpy.array(3))
 
     mock_ndim.assert_called_once_with(3)
+
+
+@pytest.mark.torch
+@pytest.mark.jax
+@pytest.mark.parametrize("array", [qml.numpy.zeros((2, 2)), torch.zeros((2, 2)), jnp.zeros((2, 2))])
+def test_set_index_jax(array):
+    """Test that an array can be created that is a copy of the
+    original array, with the value at the specified index updated"""
+
+    array2 = qml.math.set_index(array, (1, 1), 3)
+    assert qml.math.allclose(array2, np.array([[0, 0], [0, 3]]))

```

Comment:
Could you add a check that is stays the same type?

Reply:
So, currently it does not necessarily stay the same type (though it should in all these instances, so I will add a check here). 

If either `idx` or `val` are tracers, the array needs to be cast to a `jax.numpy` array. So right now, I have it converting everything to `jax` if anything is `jax` (that's also why `multi_dispatch` currently looks at all arguments).

For example, in the QMC template, we create `unitary = np.zeros((2 * M, 2 * M))` in line 128, and then update elements of it in subsequent lines. This is a `numpy` array , but if jitting, the values we update elements to are `jax` tracers. This fails, no matter what version of index setting you use, unless you cast unitary to a `jax.numpy` array.

Is this acceptable behaviour, or should this be changed? If `idx` or `val` are `jax.numpy` arrays but NOT tracers, and `array` is a standard array, then `array` will be converted to a `jax.numpy` array for no reason. I could make it dependent on `idx` and `val` being tracers instead of on which version of `numpy` they use.

Reply:
Ok I understand! I think then you have to play with the `@multi_dispatch` arguments to choose what arguments will define the interface. You can add a test for the different cases (`idx` `val` as arrays/jitted arrays). I think it is a desired behaviour that if any of the three is Jax then it defines the interface :+1: